enclave/core/sgx/calls.c

```diff
@@ -236,126 +237,225 @@ oe_result_t oe_register_ecall_function_table(
-    memcpy(input_buffer, args.input_buffer, args.input_buffer_size);
+    memcpy(enc_input_buffer, host_input_buffer, input_buffer_size);
 
     // Clear out output buffer.
     // This ensures reproducible behavior if say the function is reading from
     // output buffer.
-    output_buffer = buffer + args.input_buffer_size;
-    memset(output_buffer, 0, args.output_buffer_size);
+    enc_output_buffer = buffer + input_buffer_size;
+    memset(enc_output_buffer, 0, output_buffer_size);
 
     // Call the function.
     func(
-        input_buffer,
-        args.input_buffer_size,
-        output_buffer,
-        args.output_buffer_size,
-        &output_bytes_written);
+        enc_input_buffer,
+        input_buffer_size,
+        enc_output_buffer,
+        output_buffer_size,
+        &enc_output_bytes_written);
 
     // The output_buffer is expected to point to a marshaling struct,
     // whose first field is an oe_result_t. The function is expected
     // to fill this field with the status of the ecall.
-    result = *(oe_result_t*)output_buffer;
+    result = *(oe_result_t*)enc_output_buffer;
 
-    if (result == OE_OK)
+    if (OE_OK == result)
     {
         // Copy outputs to host memory.
-        memcpy(args.output_buffer, output_buffer, output_bytes_written);
+        memcpy(host_output_buffer, enc_output_buffer, enc_output_bytes_written);
 
         // The ecall succeeded.
-        args_ptr->output_bytes_written = output_bytes_written;
-        args_ptr->result = OE_OK;
+        *host_output_bytes_written = enc_output_bytes_written;
+        *host_result = OE_OK;
     }
 
 done:
     if (buffer)
+    {
         oe_free(buffer);
+    }
+
+    return result;
+} /* _handle_ecall_function */
 
+static oe_result_t _handle_call_enclave_function(uint64_t arg_in)
+{
+    oe_call_enclave_function_args_t args, *args_ptr;
+    oe_result_t result = OE_OK;
+
+    // Ensure that args lies outside the enclave.
+    if (!oe_is_outside_enclave(
+            (void*)arg_in, sizeof(oe_call_enclave_function_args_t)))
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Copy args to enclave memory to avoid TOCTOU issues.
+    args_ptr = (oe_call_enclave_function_args_t*)arg_in;
+    args = *args_ptr;
+
+    result = _handle_ecall_function(
+        args.table_id,
+        args.function_id,
+        (const uint8_t*)args.input_buffer,
+        args.input_buffer_size,
+        args.output_buffer,
+        args.output_buffer_size,
+        &(args_ptr->output_bytes_written),
+        &(args_ptr->result));
+
+done:
     return result;
 }
 
+static oe_result_t _handle_synchronous_switchless_ecall(
+    oe_switchless_synchronous_ecall_t* ecall)
+{
+    oe_result_t result = OE_OK;
+
+    // Ensure that args lies outside the enclave.
+    if (!oe_is_outside_enclave(
+            (void*)ecall, sizeof(oe_switchless_synchronous_ecall_t)))
+    {
+        OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    result = _handle_ecall_function(
```

Comment:
@jxyang Can you confirm if we need an lfence here? In the boundary functions, we need to ensure that there is an lfence after all the early returns have been executed.

Reply:
I don't think we need it here. Typically Spectre-1 takes advantage of speculative execution on a privileged path, e.g., OOB access into kernel memory. Here, the path can be freely executed by giving a valid `ecall` address anyway.

Reply:
I moved the __atomic_clear beyond the "done" label.  There was a potential that this would have deadlocked the host in the case that the ecall was not outside the enclave.